apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/DatastoreDbBuilder.cs

```diff
@@ -0,0 +1,247 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Api.Gax.Grpc;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Datastore.V1
+{
+    // Note: this file would be generated.
+
+    public partial class DatastoreDb
+    {
+        /// <summary>
+        /// Creates a new <see cref="DatastoreDbBuilder"/> instance. This method is primarily present for discoverability.
+        /// </summary>
+        /// <returns>A new <see cref="DatastoreDbBuilder"/> instance.</returns>
+        public DatastoreDbBuilder CreateBuilder() => new DatastoreDbBuilder();
+
+    }
+
+    /// <summary>
+    /// Builder class for <see cref="DatastoreClient"/> to provide simple configuration of credentials, endpoint etc.
+    /// </summary>
+    public sealed partial class DatastoreDbBuilder : ClientBuilderBase<DatastoreDb>
+    {
+        /// <summary>
+        /// The settings to use for RPCs, or null for the default settings.
+        /// </summary>
+        public DatastoreSettings Settings { get; set; }
+
+        private EmulatorDetection _emulatorDetection;
+
+        /// <summary>
+        /// Specifies how the builder responds to the presence of emulator environment variables as described
+        /// by https://cloud.google.com/datastore/docs/tools/datastore-emulator.
+        /// </summary>
+        /// <remarks>
+        /// This property defaults to <see cref="EmulatorDetection.None"/>, meaning that environment variables are
+        /// ignored.
+        /// </remarks>
+        public EmulatorDetection EmulatorDetection
+        {
+            get => _emulatorDetection;
+            set => _emulatorDetection = GaxPreconditions.CheckEnumValue(value, nameof(value));
+        }
+
+        /// <summary>
+        /// The project ID, or null if this has not yet been configured. Note that a project ID must be configured
+        /// (or detected via an emulator environment variable) before building the <see cref="DatastoreDb"/>.
+        /// </summary>
+        public string ProjectId { get; set; }
+
+        /// <summary>
+        /// Common code for handling the emulator configuration.
+        /// </summary>
+        private ConfiguredBuilder PrepareBuilder()
+        {
+            // TODO: This is slightly awkward, as if a new property is added to ClientBuilderBase, this code will need to be updated.
+            // An alternative would be for Build to end up setting Endpoint and ChannelCredentials on *this* object, which
+            // might also be surprising, and would prevent Build() from being called again. Hmm.
+            // Possible solution: a ClientBuilderBase constructor accepting another ClientBuilderBase, which copies all properties.
+            // We could then have an internal DatastoreClientBuilder constructor accepting a DatastoreDbBuilder.
```

Comment:
Agreed that it isn't great as it is.
Yes, I think your suggested alternative is a better approach.

Reply:
Righto, will get on that.